### ISSUE TITLE
Returned lifetimes are too restrictive.

### ISSUE THREAD

**Issue by Dirbaio** [Dirbaio/java-spaghetti#1]
If you call a method on a `Local<'env, Foo>`, it gets borrowed for a new shorter lifetime `'short`, and deref'd to `&'short Foo`. The method then returns a `Local<'short, Bar>`. 

This is unnecessarily restrictive, it could return a `Local<'env, Bar>`, Returning `Local<'short, Bar>` is annoying, for example it prevents moving or destroying the original `Foo` while the `Bar` is live.

I have prototyped a possible solution using arbitrary self types. Methods take `self: Ref<'env, Self>` instead of `&self`, so they can "see" the original `'env` lifetime and return `Local<'env>`. Seems it works nicely so far. See branch: https://github.com/Dirbaio/java-spaghetti/tree/receiver

There's a newer RFC for arbitrary self types v2, hopefully there's a push to stabilize it soon. https://github.com/rust-lang/rfcs/pull/3519

I don't know if there's a reasonably ergonomic way of solving this without arbitrary self types. Solutions are welcome.

**Comment by Dirbaio** [comment]
Updated `receiver` branch to use arbitrary self types v2. Works pretty great!

**Comment by wuwbobo2021** [comment]
~~You have pointed out that the `Deref` implementation for `Local<'env, T>` is unsound in release version 0.2.0: "this is unsound because `Ref` can be cloned/copied. You can get a Ref out of the Local, copy it, drop the Local, then use-after-free through the Ref."~~

~~It seems like your solution for this problem requires the Rust "arbitrary_self_types" feature, hopefully this will be stabilized in Rust 1.87 (I'm not sure): <https://github.com/rust-lang/rust/pull/135881>~~

**Comment by Dirbaio** [comment]
yep! I actually didn't realize it also affects the main branch. Maybe the same fix (make `Ref` not Copy/Clone, make everything take `&Ref` instead of `Ref`) should work on main, but the ObjectAndEnv hack makes it much more annoying.

**Comment by Dirbaio** [comment]
i've merged `receiver` into `main`, so this is now fixed. 

Not sure whether to wait until arbitrary_self_types is stable for a release or what

**Comment by wuwbobo2021** [comment]
Sorry, I'm relatively new to Rust, and I was threatened by `https://github.com/jni-rs/jni-rs/issues/558` when posting <https://github.com/Dirbaio/java-spaghetti/issues/1#issuecomment-2628345164>. I just checked the code again, and I cannot realize any unsound problem about `Deref`/`as_ref()` of `Local<'env, T>` in release version 0.2.0. I am safe to use this latest release version that works with the current stable compiler.

This cannot compile when using java-spaghetti 0.2.0, because `pub fn as_ref(&self) -> Ref<'_, T>` equals `pub fn as_ref<'a>(&'a self) -> Ref<'a, T>`:
```rust
                let ref1 = local.as_ref();
                let ref2 = ref1.clone();
                drop(local);
                drop(ref2);
```

This cannot compile, because `fn deref(&self) -> &Self::Target` equals `fn deref(&'a self) -> &'a Self::Target`, and it's safe because generated `ReferenceType` binding structs don't implement `Clone`:
```
                use std::ops::Deref;
                let ref_binding = local.deref();
                drop(local);
                drop(ref_binding);
```

`Deref` of `Local`:
```rust
impl<'env, T: ReferenceType> Deref for Local<'env, T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        unsafe { &*(self as *const Self as *const Self::Target) }
    }
}
```

**Comment by wuwbobo2021** [comment]
Inspired by `https://github.com/jni-rs/jni-rs/issues/558`.

I'm experimenting with my current fork of `bluest` which is based on `java-spaghetti` released version 0.2.0. I changed `AdapterImpl::default`:

```rust
        vm_context::get_vm().with_env(|env| {
            let local_context = context.as_ref(env);
            let service_name = JString::from_env_str(env, Context::BLUETOOTH_SERVICE);
            let manager = local_context
                .getSystemService_String(service_name)
                .ok()??
                .cast::<BluetoothManager>()
                .ok()?;
            let local_manager = manager.as_ref();
            let adapter = local_manager.getAdapter().ok()??;
            drop(manager);
            panic!()
        }
```

This didn't compile because `adapter` lives shorter than `manager`.

Wait a minute... please look at the generated `getAdapter` binding's signature:

```rust
        impl BluetoothManager {
            /// [getAdapter](https://developer.android.com/reference/android/bluetooth/BluetoothManager.html#getAdapter())
            pub fn getAdapter<'env>(
                &'env self,
            ) -> ::std::result::Result<
                ::std::option::Option<::java_spaghetti::Local<'env, BluetoothAdapter>>,
                ::java_spaghetti::Local<'env, super::super::java::lang::Throwable>,
            > {
```

Why is `&'env self` restriction needed? The compilation error vanished after changing it to `&self` manually. However, it's unbounded. I guess there could be an `PhantomData<Env<'env>>` stored in the Java binding type, so that `&self` wouldn't have to be `&'env self`. For `Global`, `'env` can be `'static`, but `as_local()` and `as_ref()` should be able to restrict `'env` to the real `Env<'env>`.

How about releasing a version without using Rust `arbitrary_self_types` nightly feature?

**Comment by wuwbobo2021** [comment]
Well, sorry for posting a raw idea; I've just corrected the last post.

**Comment by wuwbobo2021** [comment]
Sorry again, my last modified post is still not well-considered, it's just problematic. My current raw ideas about a few choices of solving this problem without `arbitrary_self_types`:

- Simplest: make generated binding methods require `env` as an argument in order to take the lifetime restriction. It's probably "unergonomic" to you.

- With my idea in my last post: How could `Local::as_global` even know about the type to be returned? There might be a new trait `LifeCastTo` implemented for all class binding types, so that methods of `Local`, `Global` and `Ref` that needs to return a reference of different lifetime may use this trait as a restriction of the generic return type; hopefully Rust would determine the return type automatically.

<details>

<summary>Code scratch purely for testing the possibility of such lifetime restriction</summary>

```rust
use std::marker::PhantomData;

// Simple prototype -------------------------------------------------------

trait RefType<'env> {}

impl<'env> RefType<'env> for SomeClass<'env> {}

trait LifeCastTo<'a, 'b> {
    type CastTo;
    unsafe fn life_cast_to(&self) -> &Self::CastTo;
}

/// creates `RefType<'static>` from a given `&'s RefType<'env>`.
fn new_global<'env, 's, S: RefType<'env>, T>(s: &S) -> T
    where S: LifeCastTo<'env, 'static, CastTo = T>, T: Clone,
{
    // Unsafe, simply `clone` this isn't a realistic usage of creating a global from a local!
    unsafe { s.life_cast_to().clone() }
}

// Represents a class to be generated -------------------------------------

#[derive(Clone)]
struct SomeClass<'env> {
    ph: PhantomData<&'env ()>,
}

impl<'a, 'b> LifeCastTo<'a, 'b> for SomeClass<'a> {
    type CastTo = SomeClass<'b>;
    unsafe fn life_cast_to(&self) -> &Self::CastTo {
        std::mem::transmute(self)
    }
}

// Some simple test ------------------------------------------------------

struct M;

impl M {
    fn get_some_local<'a>(&'a self) -> SomeClass<'a> {
        SomeClass { ph: PhantomData }
    }
}

fn main() {
    let man = M;
    let local = man.get_some_local();
    // compiler returns error if this is replaced with `let global = local.clone();`
    let global = new_global(&local);
    drop(man);
    drop(global);
}
```

</details>

- Another possible change: remove `Local` and `Global`, change `Ref` into an enum with `'env` but without `ReferenceType` generic parameter: `enum Ref<'env> { Local(Env<'env>, jobject), Global(jobject, PhantomData<Env<'env>>), Unowned(Env<'env>, jobject) }`. Make all class binding types to be smart pointers by themselves, owning a `ref: Ref<'env>`; add a trait `JniRef<'env>` with a required method `fn jni_ref(&self) -> &Ref<'env>;` which provides some methods originally provided by `Local`, `Global`, or `Ref` (some other methods still have to be generated in the bindings).

**Comment by wuwbobo2021** [comment]
The 3rd raw idea in <https://github.com/Dirbaio/java-spaghetti/issues/1#issuecomment-3066688145> is a bad idea, I realized that `Deref` would not be possible for these class bindings because of the `Global` in the imaginary new `Ref` enum: the global reference type may look like `SomeClass<'static>`, but how to limit it to `'env` when dereferencing it? Not possible. To make it safe, I might make direct dereferencing of `SomeClass` panic if it's a global, but it's quite erroneous.

Note for everyone to realize the inconvenience and performance cost of the problem "returned lifetimes are too restrictive" quickly: a function that receives a local reference cannot return a local reference belonging to the same `JNIEnv`, it must be turned into a global before returning.

**Comment by wuwbobo2021** [comment]
I'm really sorry that I have probably led you to make the decision of merging the nightly branch (relying on the nightly Rust feature) into the main branch. I'm here to remind you again that my comments in <https://github.com/Dirbaio/java-spaghetti/issues/1#issuecomment-2628345164> are absolutely wrong.
